pkg/controller/case_test.go

```diff
@@ -0,0 +1,589 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller_test
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"testing"
+	"time"
+
+	"github.com/kubernetes-incubator/service-catalog/pkg/apis/servicecatalog/v1beta1"
+	fakesc "github.com/kubernetes-incubator/service-catalog/pkg/client/clientset_generated/clientset/fake"
+	scinterface "github.com/kubernetes-incubator/service-catalog/pkg/client/clientset_generated/clientset/typed/servicecatalog/v1beta1"
+	scinformers "github.com/kubernetes-incubator/service-catalog/pkg/client/informers_generated/externalversions"
+	"github.com/kubernetes-incubator/service-catalog/pkg/controller"
+	"github.com/pmorie/go-open-service-broker-client/v2"
+	osb "github.com/pmorie/go-open-service-broker-client/v2"
+	fakeosb "github.com/pmorie/go-open-service-broker-client/v2/fake"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	k8sinformers "k8s.io/client-go/informers"
+	fakek8s "k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/tools/record"
+)
+
+const (
+	testNamespace                         = "test-ns"
+	testClusterServiceBrokerName          = "test-clusterservicebroker"
+	testClusterServiceClassName           = "test-clusterserviceclass"
+	testClusterServicePlanName            = "test-clusterserviceplan"
+	testServiceInstanceName               = "service-instance"
+	testClassExternalID                   = "clusterserviceclass-12345"
+	testPlanExternalID                    = "34567"
+	testNonbindablePlanExternalID         = "nb34567"
+	testNonbindableClusterServicePlanName = "test-nonbindable-plan"
+	testExternalID                        = "9737b6ed-ca95-4439-8219-c53fcad118ab"
+	testBindingName                       = "test-binding"
+	testServiceBindingGUID                = "bguid"
+	authSecretName                        = "basic-secret-name"
+
+	pollingInterval = 50 * time.Millisecond
+	pollingTimeout  = 8 * time.Second
+)
+
+type ControllerTest struct {
+	// resource clientsets and interfaces
+	scInterface      scinterface.ServicecatalogV1beta1Interface
+	k8sClient        *fakek8s.Clientset
+	fakeOSBClient    *fakeosb.FakeClient
+	catalogReactions []fakeosb.CatalogReaction
+	osbClientCfg     *v2.ClientConfiguration
+	stopCh           chan struct{}
+}
+
+func NewControllerTest(t *testing.T) *ControllerTest {
```

Comment:
I think the helper methods in here should be commented as well

Reply:
we do not need to have it public, I've renamed it to newControllerTest

Reply:
I mean, i would still want them commented so someone new coming in can know what they do and how to use them

Reply:
ok, I put comments